File: Patterns/Command Pattern/Cpp/Command/Stereo/StereoOnWithCdCommand.cpp
#include "../ICommand.cpp"
#include "./Stereo.cpp"

class StereoOnWithCdCommand : public ICommand {
private:
    Stereo* stereo;
    int preview;

public:
    StereoOnWithCdCommand(Stereo* stereo) : stereo{stereo}, preview{stereo->OFF} {}

    void execute() override {
        preview = stereo->getState();
        stereo->on();
        stereo->setCd();
        stereo->setVolumn(11);
    }

    void undo() override {
        if (preview == stereo->DVD) stereo->setDvd();
        else if (preview == stereo->RADIO) stereo->setRadio();
        else if (preview == stereo->OFF) stereo->off();
    }

    string getCommandName() override {
        return "StereoOnWithCdCommand";
    }
};

File: Patterns/Command Pattern/CSharp/Command/Stereo/StereoOnWithCdCommand.cs
namespace Command.Stereo
{
    public class StereoOnWithCdCommand : ICommand
    {
        private Stereo stereo;
        private int preview;

        public StereoOnWithCdCommand(Stereo stereo)
        {
            preview = Stereo.OFF;
            this.stereo = stereo;
        }

        public void Execute()
        {
            preview = stereo.State;
            stereo.On();
            stereo.SetCd();
            stereo.SetVolumn(11);
        }

        public void Undo()
        {
            switch (preview)
            {
                case Stereo.DVD:
                    stereo.SetDvd();
                    break;
                case Stereo.RADIO:
                    stereo.SetDvd();
                    break;
                case Stereo.OFF:
                    stereo.Off();
                    break;
            }
        }
    }
}